### PR TITLE
Refactor/pep

### PR DESCRIPTION
**Pull request by bgoujaud**
L ensemble des changements est décrit en messages de commits:
Tout est fait dans PEP.
Le reste concerne l utilisation des changements.
J ai vérifié:
- les tests
- les examples
- la doc
- le readme
- le notebook
- le papier

Mais si vous voulez verifier que je n ai rien n oublié, c est cool! ;)

### CODE REVIEW

**Review by AdrienTaylor — CHANGES_REQUESTED** [review]
+ error in the test (I don't get it):
==============================================================
>> python -m unittest
(PEPit) Setting up the problem: size of the main PSD matrix: 4x4
(PEPit) Setting up the problem: performance measure is minimum of 1 element(s)
(PEPit) Setting up the problem: initial conditions (2 constraint(s) added)
(PEPit) Setting up the problem: interpolation conditions for 1 function(s)
		 function 1 : 6 constraint(s) added
(PEPit) Compiling SDP
(PEPit) Calling SDP solver
(PEPit) Solver status: optimal (solver: MOSEK); optimal value: -6.331018474007843e-14
(PEPit) Postprocessing: solver's output is not entirely feasible (smallest eigenvalue of the Gram matrix is: -6.88e-14 < 0).
 Small deviation from 0 may simply be due to numerical error. Big ones should be deeply investigated.
 In any case, from now the provided values of parameters are based on the projection of the Gram matrix onto the cone of symmetric semi-definite matrix.
F..............................................................................................
======================================================================
FAIL: test_counter (tests.test_constraints.TestConstraints)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/adrien/Github/PEPit/PEPit/tests/test_constraints.py", line 67, in test_counter
    self.assertIs(self.problem.list_of_constraints[i].counter, i)
AssertionError: 8 is not 0

----------------------------------------------------------------------
Ran 95 tests in 15.711s

FAILED (failures=1)
======================================================================
>>python -m unittest tests.test_constraints
.....
----------------------------------------------------------------------
Ran 5 tests in 0.152s

OK

File: PEPit/pep.py [hunk 1]
```diff
@@ -173,24 +203,40 @@ def _expression_to_cvxpy(expression, F, G):
...
-                               the dimension of the solution (rank of the Gram matrix).
+            verbose (int): Level of information details to print (Override the CVXPY solver verbose parameter).
+                           0: No verbose at all
+                           1: PEPit information is printed but not CVXPY's
+                           2: Both PEPit and CVXPY details are printed
             return_full_cvxpy_problem (bool): If True, return the cvxpy Problem object.
                                               If False, return the worst case value only.
                                               Set to False by default.
+            dimension_reduction (bool): Minimize the dimension of the solution (rank of the Gram matrix).
+            dimension_reduction_heuristic (str, optional): The heuristic to be minimized to reduce the dimension.
```
Comment (AdrienTaylor) [inline PEPit/pep.py:222]: An heuristic to reduce the dimension

**Review by AdrienTaylor — CHANGES_REQUESTED** [review]

**Review by AdrienTaylor — APPROVED** [review]

**Comment by M-Celine** [comment]
> * error in the test (I don't get it):
>   ==============================================================
> 
> > > python -m unittest
> > > (PEPit) Setting up the problem: size of the main PSD matrix: 4x4
> > > (PEPit) Setting up the problem: performance measure is minimum of 1 element(s)
> > > (PEPit) Setting up the problem: initial conditions (2 constraint(s) added)
> > > (PEPit) Setting up the problem: interpolation conditions for 1 function(s)
> > > function 1 : 6 constraint(s) added
> > > (PEPit) Compiling SDP
> > > (PEPit) Calling SDP solver
> > > (PEPit) Solver status: optimal (solver: MOSEK); optimal value: -6.331018474007843e-14
> > > (PEPit) Postprocessing: solver's output is not entirely feasible (smallest eigenvalue of the Gram matrix is: -6.88e-14 < 0).
> > > Small deviation from 0 may simply be due to numerical error. Big ones should be deeply investigated.
> > > In any case, from now the provided values of parameters are based on the projection of the Gram matrix onto the cone of symmetric semi-definite matrix.
> > > F..............................................................................................
> > > ======================================================================
> > > FAIL: test_counter (tests.test_constraints.TestConstraints)
> 
> Traceback (most recent call last): File "/home/adrien/Github/PEPit/PEPit/tests/test_constraints.py", line 67, in test_counter self.assertIs(self.problem.list_of_constraints[i].counter, i) AssertionError: 8 is not 0
> 
> Ran 95 tests in 15.711s
> 
> # FAILED (failures=1)
> > > python -m unittest tests.test_constraints
> > > .....
> 
> Ran 5 tests in 0.152s
> 
> OK

I don't have this problem

**Review by M-Celine — APPROVED** [review]
Ok for me, thanks! 
I have with some remarks/questions : 
* You modified set_initial_condition in pep.py to call add_constraints, and then suppress all the call to this function in the examples. Do we really need to keep it then?
* Do we really need two arguments for dimension reduction (dimension_reduction and dimension_reduction_heuristic)?
* In the QuickStart, add dimension_reduction_heuristic="trace" when calling problem.solve(dimension_reduction=True).

**Pull request merged**